javascript/ql/src/semmle/javascript/Promises.qll

```diff
@@ -1,9 +1,295 @@
+ * A resolved promise created by the standard ECMAScript 2015 `Promise.resolve` function.
+ */
+class ResolvedES2015PromiseDefinition extends ResolvedPromiseDefinition {
+  ResolvedES2015PromiseDefinition() {
+    this = DataFlow::globalVarRef("Promise").getAMemberCall("resolve")
+  }
+
+  override DataFlow::Node getValue() { result = getArgument(0) }
+}
+
+/**
+ * An aggregated promise produced either by `Promise.all` or `Promise.race`.
+ */
+class AggregateES2015PromiseDefinition extends PromiseCreationCall {
+  AggregateES2015PromiseDefinition() {
+    exists(string m | m = "all" or m = "race" |
+      this = DataFlow::globalVarRef("Promise").getAMemberCall(m)
+    )
+  }
+
+  override DataFlow::Node getValue() {
+    result = getArgument(0).getALocalSource().(DataFlow::ArrayCreationNode).getAnElement()
+  }
+}
+
+/**
+ * This module defines how exceptional data-flow propagates into and out a Promise.
+ */
+private module ExceptionalPromiseFlow {
+  /**
+   * Gets the pseudo-field used to describe rejected values in a promise.
+   */
+  string rejectField() {
+    result = "$PromiseRejectField$"
+  }
+  
+  /**
+   * A flow step describing a promise definition.
+   *
+   * The rejected value is written to a pseudo-field on the promise.
+   */
+  class PromiseDefitionStep extends DataFlow::AdditionalFlowStep {
+    PromiseDefinition promise;
+    PromiseDefitionStep() {
+      this = promise
+    }
+
+    override predicate store(DataFlow::Node pred, DataFlow::Node succ, string prop) {
+      prop = rejectField() and
+      (
+        pred = promise.getRejectParameter().getACall().getArgument(0) or
+        pred = promise.getExecutor().getExceptionalReturn()
+      ) and
+      succ = this
+    }
+  }
+  
+  /**
+   * A load step loading the pseudo-field describing that the promise is rejected.
+   * The rejected value is thrown as a exception.
+   */
+  class AwaitStep extends DataFlow::AdditionalFlowStep {
+    DataFlow::Node operand;
+    AwaitExpr await;
+    AwaitStep() {
+      this.getEnclosingExpr() = await and
+      operand.getEnclosingExpr() = await.getOperand()
+    }
+
+    override predicate load(DataFlow::Node pred, DataFlow::Node succ, string prop) {
+      prop = rejectField() and
+      succ = await.getExceptionTarget() and
+      pred = operand
+    }
+  }
+
+  /**
+   * A flow step describing the data-flow related to the `.then` method of a promise.
+   */
+  class ThenStep extends DataFlow::AdditionalFlowStep, DataFlow::MethodCallNode {
+    ThenStep() {
+      this.getMethodName() = "then"
+    }
+
+    override predicate load(DataFlow::Node pred, DataFlow::Node succ, string prop) {
+      prop = rejectField() and
+      pred = getReceiver() and
+      succ = getCallback(1).getParameter(0)
+    }
+    
+    override predicate copyProperty(DataFlow::Node pred, DataFlow::Node succ, string prop) {
+      not exists(this.getArgument(1)) and
+      prop = rejectField() and
+      pred = getReceiver() and
+      succ = this
```

Comment:
Do we handle the case where callback(0) returns a failed promise? It seems like there should be a step for that here

Reply:
No, we do not.   
That should be pretty easy to model using `copyProperty()`. 

Same with callbacks that return resolved promises.

Reply:
Oh wait, I said callback(0), but actually it should be `getCallback([0..1])` like you've done a few other places as well.